File: python/src/operations/pix_operation.py
from src.entities.person import Person
from src.interfaces.iaccountmethod import IOperationMethod
from src.utils.general import PIXInfo

class PIX(IOperationMethod):
    def do_transaction(self, sender: Person, recipient: Person, amount: float):
        print(PIXInfo.START_PIX.value)

        try:
            if self.validate_transaction(sender, recipient, amount):
                sender.account.update_balance(amount)
                recipient.account.update_balance(amount)
            
        except Exception as e:
            return e

    def validate_transaction(self, sender: Person, recipient: Person, amount: float):
        return recipient.account.get_details().owner == recipient.identity() and sender.account.get_details().owner == sender.identity() and amount != 0.0

File: go/pkg/operations/pix_operation.go
package operations

import (
	"bank-system/pkg/entities"
	"fmt"
	"log"
)

type PIX struct{}

func (t *PIX) DoTransaction(sender, recipient entities.IPerson, amount float32) (string, error) {
	log.Println("Starting PIX transaction")
	if t.validateTransaction(sender, recipient, amount) {
		senderAcc, err := sender.GetAccount()
		if err != nil {
			return "Cannot do transaction", fmt.Errorf("cannot do transaction")
		}
		senderAcc.UpdateBalance(amount)
		recipAcc, err := recipient.GetAccount()
		if err != nil {
			return "Cannot do transaction", fmt.Errorf("cannot do transaction")
		}
		recipAcc.UpdateBalance((amount * -1))
		return "PIX transaction completed", nil
	}
	return "Error", fmt.Errorf("Error")
}

func (t *PIX) validateTransaction(sender, recipient entities.IPerson, amount float32) bool {
	senderacc, err := sender.GetAccount()
	if err != nil {
		return false
	}
	recipientAcc, err := recipient.GetAccount()
	if err != nil {
		return false
	}
	return senderacc.GetOwner() == sender.GetIdentity() && recipientAcc.GetOwner() == recipient.GetIdentity() && amount != 0
}
